conmon-rs/server/src/child_reaper.rs

```diff
@@ -1,169 +1,148 @@
+use anyhow::{format_err, Context, Result};
 use log::{debug, error};
+use multimap::MultiMap;
 use nix::sys::signal::{kill, Signal};
 use nix::sys::wait::{waitpid, WaitStatus};
 use nix::unistd::Pid;
 use std::path::PathBuf;
-use std::thread;
-use std::{collections::HashMap, fs::File, io::Write, sync::Arc, sync::Mutex};
-
-impl ChildReaper {
-    pub fn init_reaper(&self) -> Result<()> {
-        let grandchildren = Arc::clone(&self.grandchildren);
-        let wait_lock = Arc::clone(&self.wait_lock);
-        thread::spawn(move || {
-            loop {
-                // To prevent an error: "No child processes (os error 10)" when spawning new runtime processes,
-                // we don't call waitpid until the runtime process has finished
-                // Inspired by
-                // https://github.com/kata-containers/kata-containers/blob/828a304883a4/src/agent/src/signal.rs#L27
-                let wait_status = {
-                    let _lock = wait_lock.lock();
-                    waitpid(Pid::from_raw(-1), None)
-                };
+use std::sync::Mutex;
+use std::{fs::File, io::Write, sync::Arc};
 
-                match wait_status {
-                    Ok(status) => {
-                        if let WaitStatus::Exited(grandchild_pid, exit_status) = status {
-                            let grandchildren = Arc::clone(&grandchildren);
-                            thread::spawn(move || {
-                                if let Err(e) = Self::forget_grandchild(
-                                    grandchildren,
-                                    grandchild_pid,
-                                    exit_status,
-                                ) {
-                                    error!(
-                                        "Failed to reap grandchild for pid {}: {}",
-                                        grandchild_pid, e
-                                    );
-                                }
-                            });
-                        }
-                    }
+#[derive(Debug, Default)]
+pub struct ChildReaper {
+    grandchildren: Arc<Mutex<MultiMap<String, ReapableChild>>>,
+}
 
-                    Err(err) => {
-                        // TODO FIXME this busy loops right now while there are no grandchildren.
-                        // There should be a broadcast mechanism so we only run this loop
-                        // while there are grandchildren
-                        if err != nix::errno::Errno::ECHILD {
-                            error!("caught error in reading for sigchld {}", err);
-                        }
-                    }
-                }
-            }
-        });
-        Ok(())
+impl ChildReaper {
+    pub fn get(&self, id: String) -> Result<ReapableChild> {
+        let locked_grandchildren = Arc::clone(&self.grandchildren);
+        let lock = locked_grandchildren.lock().unwrap();
+        let r = lock.get(&id).context("")?.clone();
+        drop(lock);
```

Comment:
do we need to manually call drop here? I would expect it to go out of scope after the Ok

Reply:
yes, the drop needs to be called to allow the borrow checker to succeed.